src/util/simplify_expr.cpp

```diff
@@ -2402,6 +2416,7 @@ bool simplify_exprt::simplify_rec(exprt &expr)
 
   if(!result)
   {
+    POSTCONDITION(tmp.type() == expr.type());
```

Comment:
We are hitting this invariant - will try to produce a jbmc version of the crash at some point today.

Reply:
Ok, that's  another bug in the simplifier then or some weird expression being passed in. Some of the bug fixes already in this PR are a consequence of adding this invariant.

Reply:
The problem is we've simplified something that has type `pointer_typet(symbol_typet(java.lang.Object))` and end up with `pointer_typet(struct_typet(.tag = java.lang.Object)`. 

I've never really be clear whether this transformation (`symbol_typet` -> `ns.follow(symbol_type)`) should always be ignored? The bug relating to multi-dimensional arrays seems to arise from the fact the element type of the array is stored on the `symbol_typet` and is lost if it is `ns.follow`'d. Will see if happens on JBMC

Reply:
In this PR I've made a few changes to the simplifier to restore the original type when `type_before != type_after` despite `base_type_eq(type_before, type_after, ns)` (which is exactly the case for `symbol_typet` expansion. In my view, the simplifier should not change the type at all and therefore should be careful to restore the symbol type where needed. We could of course, instead, relax the `POSTCONDITION` to check for semantic equivalence (via `base_type_eq`) rather than syntactic equivalence, but then this would negatively affect examples like the multi-dimensional array that you mentioned.

So what it would take is tracking down what simplification is being performed and then fix this in the same way it's already being done several times in this PR.

Reply:
I can't reproduce this yet on JBMC as failing example seems to depend on a model for `java.lang.Class`. It seems like `util/type_eq` should ignore symbolically equivalent types and could be used here, but it doesn't recurse unfortunately. Not got more time to investigate today but can have a bit more of a prod tomorrow.

Reply:
In order to debug this you will want to print `expr.id()` before this failing `POSTCONDITON` (and possibly `#define DEBUGX`). That should suffice to know which simplification rule to look at.

Reply:
Found it - tautschnig/cbmc#7 - will verify fixes other issues. You might have some insight on how to create a regression that demonstrates this on the CBMC side?

Reply:
A test should look like this, except it doesn't really do the job right now, because the update to `x` is not constant-propagated.
```
#include <assert.h>

struct A
{
  struct A *next;
};

int main()
{
  struct A x, y;
  x.next = &y;
  assert(x.next == &y);
  return 0;
}
```

Reply:
There appears to be more - will investigate these as well.

Reply:
Ignore previous, appears to be PEBKAC. TG pointer updated - will post if passed. 

Re regression - in our previously broken example, it is simplified to a null pointer, so perhaps assigning `x.next = NULL` will work? Will quickly give it a go

Reply:
Nope - works before and after using `cbmc test.c`